{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# PhiSpy\n",
    "\n",
    "This is a Jupyter Notebook that shows how to run PhiSpy manually. You can run through all the steps that PhiSpy takes to determine whether a genome contains a prophage, and inspect all of the data generated by PhiSpy.\n",
    "\n",
    "You will need to install [PhiSpy](https://github.com/linsalrob/PhiSpy#installation), [Jupyter Notebooks](https://jupyter.org/install), and [pandas](https://pandas.pydata.org/pandas-docs/stable/getting_started/install.html)\n",
    "\n",
    "<small>Note: PhiSpy does not normally use pandas, but we use it here to visualize the data!</small>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# set up the environment\n",
    "import os\n",
    "import sys\n",
    "import gzip\n",
    "from functools import reduce\n",
    "import tempfile\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "from Bio import SeqIO\n",
    "\n",
    "import PhiSpyModules"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Check the PhiSpy version\n",
    "\n",
    "We recommend at least version 4.0.3 but preferable 4.1 or higher"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PhiSpy version: 4.1.10\n"
     ]
    }
   ],
   "source": [
    "print(\"PhiSpy version: \" + PhiSpyModules.__version__)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Define your genbank file here.\n",
    "\n",
    "You may need to set the full path to the file. You can use gzip compressed file or an uncompressed file. Obviously, you will know whether it is compressed or not, but this demonstrates how PhiSpy determines when a file is compressed.\n",
    "\n",
    "**This should be the only line you need to change to run PhiSpy completely!**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "genbankfile = \"/home/redwards/GitHubs/PhiSpy/test_genbank_files/Streptococcus_pyogenes_M1_GAS.gb.gz\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Parse the file\n",
    "\n",
    "We use BioPython to parse the file, but we also add a few additional mehtods to the standard BioPython object to ease parsing. We also merge or split compound features (those with more than one location along the chromosome) to appropriately handle them.\n",
    "\n",
    "Our `record` object is an extended `SeqIO.parse` object."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 1 contigs to predict prophages on!\n"
     ]
    }
   ],
   "source": [
    "min_contig_size = 1000\n",
    "\n",
    "if PhiSpyModules.is_gzip_file(genbankfile):\n",
    "    handle = gzip.open(genbankfile, 'rt')\n",
    "else:\n",
    "    handle = open(genbankfile, 'r')\n",
    "\n",
    "record  = PhiSpyModules.SeqioFilter(filter(lambda x: len(x.seq) > min_contig_size, SeqIO.parse(handle, \"genbank\")))\n",
    "handle.close()\n",
    "\n",
    "# we check to make sure there are some contigs left to process\n",
    "ncontigs = reduce(lambda sum, element: sum + 1, record, 0)\n",
    "print(f\"There are {ncontigs} contigs to predict prophages on!\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Define the parameters that we will use\n",
    "\n",
    "These are normally provided as command line options, but for jupyter we set them here\n",
    "\n",
    "paramter | meaning | options | default value\n",
    "--- | --- | --- | --\n",
    "kmers_type | What do we count kmers with? | `all`, `codon`, `simple` | `all`\n",
    "window_size | How many consecutive ORFs to include? | an integer | 30\n",
    "record | the `Bio.SeqIO` object with all the sequences | a `Bio.SeqIO` object | `record`\n",
    "expand_slope | whether to use the square of the slope of the Shannon scores | `True` or `False` | `False`\n",
    "number | Number of consecutive genes in a region of window size that must be prophage genes | an integer | 5\n",
    "nonprophage_genegaps | The number of non phage genes betweeen prophages | an integer | 10\n",
    "quiet | Don't make additional outputs | `True` or `False` | `True`\n",
    "\n",
    "\n",
    "*Note*: You can add an additional paramter, `make_training_data` here (its actual value doesn't matter) that will append an additional column to the output for each ORF that includes a `1` (`True`) if the ORF is thought or stated to be a phage gene or `0` (`False`) otherwise."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "parameters = {\n",
    "    'kmers_type': \"all\",\n",
    "    'window_size': 30,\n",
    "    'record': record,\n",
    "    'expand_slope': False,\n",
    "    'training_set': \"data/trainSet_genericAll.txt\",\n",
    "    'randomforest_trees': 500,\n",
    "    'threads': 4,\n",
    "    'quiet': True,\n",
    "    'nonprophage_genegaps': 10,\n",
    "    'number': 5,\n",
    "    'color' : True,\n",
    "    'debugpp' : 'all_pp_jup.tsv',\n",
    "    'evaluate':False,\n",
    "    'make_training_data':None,\n",
    "    'skip_search':False,\n",
    "    'phmms':None,\n",
    "    'phage_genes' : 2,\n",
    "    'metrics' : ['orf_length_med', 'shannon_slope', 'at_skew', 'gc_skew', 'max_direction'],\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Generate the test data\n",
    "\n",
    "This is the step that actually does all the measurements!\n",
    "\n",
    "In this example, we convert the output to a pandas dataframe for visualization and exploration."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>orf_length_med</th>\n",
       "      <th>shannon_slope</th>\n",
       "      <th>at_skew</th>\n",
       "      <th>gc_skew</th>\n",
       "      <th>max_direction</th>\n",
       "      <th>phmms</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>393.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.305653</td>\n",
       "      <td>1.598561</td>\n",
       "      <td>14</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>382.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.379462</td>\n",
       "      <td>1.520969</td>\n",
       "      <td>15</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>372.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.108348</td>\n",
       "      <td>1.521077</td>\n",
       "      <td>16</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>295.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.913390</td>\n",
       "      <td>1.503681</td>\n",
       "      <td>17</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>372.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.886275</td>\n",
       "      <td>1.561022</td>\n",
       "      <td>18</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   orf_length_med  shannon_slope   at_skew   gc_skew  max_direction  phmms\n",
       "0           393.0            0.0  2.305653  1.598561             14    0.0\n",
       "1           382.5            0.0  2.379462  1.520969             15    0.0\n",
       "2           372.0            0.0  2.108348  1.521077             16    0.0\n",
       "3           295.5            0.0  1.913390  1.503681             17    0.0\n",
       "4           372.0            0.0  1.886275  1.561022             18    0.0"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "parameters['test_data'] = PhiSpyModules.measure_features(**parameters)\n",
    "test_data = parameters['test_data']\n",
    "# note that if you include make_training_data you will need to add an \"is_phage\" column here\n",
    "test_df = pd.DataFrame(parameters['test_data'], columns = ['orf_length_med', 'shannon_slope', 'at_skew',\n",
    "                                             'gc_skew', 'max_direction', 'phmms'])\n",
    "test_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Run the random forest\n",
    "\n",
    "Here we run the random forest to identify the phages, and combine that into our initial table as the `rank` column."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene id</th>\n",
       "      <th>function</th>\n",
       "      <th>contig</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>position</th>\n",
       "      <th>rank</th>\n",
       "      <th>my status</th>\n",
       "      <th>pp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>[231:1587](+)</td>\n",
       "      <td>Chromosomal replication initiator protein DnaA</td>\n",
       "      <td>NC_002737</td>\n",
       "      <td>232</td>\n",
       "      <td>1587</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>[1741:2878](+)</td>\n",
       "      <td>DNA polymerase III beta subunit (EC 2.7.7.7)</td>\n",
       "      <td>NC_002737</td>\n",
       "      <td>1742</td>\n",
       "      <td>2878</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>[2952:3150](+)</td>\n",
       "      <td>Uncharacterized protein CAC3725</td>\n",
       "      <td>NC_002737</td>\n",
       "      <td>2953</td>\n",
       "      <td>3150</td>\n",
       "      <td>2</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>[3295:3415](-)</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>NC_002737</td>\n",
       "      <td>3415</td>\n",
       "      <td>3296</td>\n",
       "      <td>3</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>[3479:4595](+)</td>\n",
       "      <td>GTP-binding and nucleic acid-binding protein YchF</td>\n",
       "      <td>NC_002737</td>\n",
       "      <td>3480</td>\n",
       "      <td>4595</td>\n",
       "      <td>4</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          gene id                                           function  \\\n",
       "0   [231:1587](+)     Chromosomal replication initiator protein DnaA   \n",
       "1  [1741:2878](+)       DNA polymerase III beta subunit (EC 2.7.7.7)   \n",
       "2  [2952:3150](+)                    Uncharacterized protein CAC3725   \n",
       "3  [3295:3415](-)                               hypothetical protein   \n",
       "4  [3479:4595](+)  GTP-binding and nucleic acid-binding protein YchF   \n",
       "\n",
       "      contig  start  stop  position    rank  my status   pp  \n",
       "0  NC_002737    232  1587         0  0.0000          0  0.0  \n",
       "1  NC_002737   1742  2878         1  0.0000          0  0.0  \n",
       "2  NC_002737   2953  3150         2  0.0000          0  0.5  \n",
       "3  NC_002737   3415  3296         3  0.0000          0  0.5  \n",
       "4  NC_002737   3480  4595         4  0.0001          0  0.0  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "parameters['rfdata'] = PhiSpyModules.call_randomforest(**parameters)\n",
    "parameters['initial_tbl'] = PhiSpyModules.make_initial_tbl(**parameters)\n",
    "parameters['output_dir'] = tempfile.mkdtemp()\n",
    "initial_table_df = pd.DataFrame(parameters['initial_tbl'], columns = ['gene id', 'function', 'contig', 'start', 'stop', 'position', 'rank', 'my status', 'pp'])\n",
    "initial_table_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1853"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(parameters['test_data'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1853,)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "parameters['rfdata'].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "numpy.ndarray"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "type(parameters['rfdata'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1853,)"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.zeros(len(parameters['test_data'])).shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1853"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(parameters['test_data'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "from io import TextIOWrapper\n",
    "from sklearn.cluster import KMeans\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "import numpy as np\n",
    "\n",
    "def call_randomforest_local(**kwargs):\n",
    "    training_file = kwargs['training_set_full']\n",
    "    test_data = kwargs['test_data']\n",
    "\n",
    "    if not os.path.exists(training_file):\n",
    "        sys.stderr.write(f\"Training file {training_file} not found. Can't continue\\n\")\n",
    "        return None\n",
    "    \n",
    "    strm = open(training_file, 'rb')\n",
    "    train_data = np.genfromtxt(TextIOWrapper(strm), delimiter=\"\\t\", skip_header=1, filling_values=1)\n",
    "\n",
    "    all_metrics = ['orf_length_med', 'shannon_slope', 'at_skew', 'gc_skew', 'max_direction', 'phmms']\n",
    "    if kwargs['phmms']:\n",
    "        kwargs['metrics'].append('phmms')\n",
    "\n",
    "    if len(kwargs['metrics']) < len(all_metrics):\n",
    "        sys.stderr.write(f\"Using the following metric(s): {', '.join(sorted(kwargs['metrics']))}.\\n\")\n",
    "        skip_metrics = [all_metrics.index(x) for x in set(all_metrics) - set(kwargs['metrics'])]\n",
    "        train_data = np.delete(train_data, skip_metrics, 1)\n",
    "        test_data = np.delete(test_data, skip_metrics, 1)\n",
    "    else:\n",
    "        sys.stderr.write(f\"Using all metrics: {', '.join(all_metrics)}.\\n\")\n",
    "\n",
    "    \"\"\"\n",
    "    Przemek's comment\n",
    "    by default 10 until version 0.22 where default is 100\n",
    "    number of estimators also implies the precision of probabilities, generally 1/n_estimators\n",
    "    in R's randomForest it's 500 and the usage note regarding number of trees to grow says:\n",
    "    \"This should not be set to too small a number, to ensure that every input row gets predicted at least a few times.\"\n",
    "    \"\"\"\n",
    "    sys.stderr.write(f\"Train data is {train_data.shape}\\n\")\n",
    "    if train_data.shape[1] > 1:\n",
    "        clf = RandomForestClassifier(n_estimators=kwargs['randomforest_trees'], n_jobs=kwargs['threads'])\n",
    "        clf.fit(train_data[:, :-1], train_data[:, -1].astype('int'))\n",
    "        return clf.predict_proba(test_data)[:,1]\n",
    "    else:\n",
    "        return np.zeros(len(kwargs['test_data']))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "parameters['test_data'] = test_data\n",
    "parameters['training_set_full'] = '/home/redwards/GitHubs/PhiSpy/PhiSpyModules/data/trainSet_genericAll.txt'\n",
    "#parameters['metrics'] = [\"xxxx\"]\n",
    "parameters['metrics'] = ['orf_length_med', 'shannon_slope', 'at_skew', 'gc_skew', 'max_direction', 'phmms']\n",
    "\n",
    "clf = call_randomforest_local(**parameters)\n",
    "clf.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "Number of features of the model must match the input. Model n_features is 5 and input n_features is 6 ",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-35-3be466cc02c8>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mclf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpredict_proba\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtest_data\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0mr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshape\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/lib/python3.8/dist-packages/scikit_learn-0.23.1-py3.8-linux-x86_64.egg/sklearn/ensemble/_forest.py\u001b[0m in \u001b[0;36mpredict_proba\u001b[0;34m(self, X)\u001b[0m\n\u001b[1;32m    671\u001b[0m         \u001b[0mcheck_is_fitted\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    672\u001b[0m         \u001b[0;31m# Check data\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 673\u001b[0;31m         \u001b[0mX\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_validate_X_predict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mX\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    674\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    675\u001b[0m         \u001b[0;31m# Assign chunk of trees to jobs\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/lib/python3.8/dist-packages/scikit_learn-0.23.1-py3.8-linux-x86_64.egg/sklearn/ensemble/_forest.py\u001b[0m in \u001b[0;36m_validate_X_predict\u001b[0;34m(self, X)\u001b[0m\n\u001b[1;32m    419\u001b[0m         \u001b[0mcheck_is_fitted\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    420\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 421\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mestimators_\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_validate_X_predict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mX\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcheck_input\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    422\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    423\u001b[0m     \u001b[0;34m@\u001b[0m\u001b[0mproperty\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/lib/python3.8/dist-packages/scikit_learn-0.23.1-py3.8-linux-x86_64.egg/sklearn/tree/_classes.py\u001b[0m in \u001b[0;36m_validate_X_predict\u001b[0;34m(self, X, check_input)\u001b[0m\n\u001b[1;32m    394\u001b[0m         \u001b[0mn_features\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mX\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshape\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    395\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mn_features_\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0mn_features\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 396\u001b[0;31m             raise ValueError(\"Number of features of the model must \"\n\u001b[0m\u001b[1;32m    397\u001b[0m                              \u001b[0;34m\"match the input. Model n_features is %s and \"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    398\u001b[0m                              \u001b[0;34m\"input n_features is %s \"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: Number of features of the model must match the input. Model n_features is 5 and input n_features is 6 "
     ]
    }
   ],
   "source": [
    "r = clf.predict_proba(test_data)\n",
    "r.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Refine the predictions\n",
    "\n",
    "Finally, we refine the predictions from the random forest and other metrics, and then predict the *att* sequences."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "parameters['pp'] = PhiSpyModules.fixing_start_end(**parameters)\n",
    "pp_df = pd.DataFrame.from_dict(parameters['pp']).transpose()\n",
    "pp_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Our `pp_df` data frame has our final prophage predictions for this genome! "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Make the final table\n",
    "\n",
    "Here we just append the pp number of the prophage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "parameters['final_tbl'] = []\n",
    "for i in parameters['initial_tbl']:\n",
    "    my_fs = PhiSpyModules.evaluation.check_pp(i[2], i[3], i[4], parameters['pp'])\n",
    "    parameters['final_tbl'].append(i + [my_fs])\n",
    "\n",
    "    final_df  = pd.DataFrame(parameters['final_tbl'], columns = ['gene id', 'function', 'contig', 'start', 'stop', 'position', 'rank', 'my status', 'pp', 'final status'])\n",
    "final_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "phage_regions = final_df.loc[final_df['final status'] > 0]\n",
    "phage_regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rp = initial_table_df.loc[initial_table_df['function'].str.contains(pat = 'ribosomal protein')]\n",
    "rp"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
