{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# PhiSpy\n",
    "\n",
    "This is a Jupyter Notebook that shows how to run PhiSpy manually. You can run through all the steps that PhiSpy takes to determine whether a genome contains a prophage, and inspect all of the data generated by PhiSpy.\n",
    "\n",
    "You will need to install [PhiSpy](https://github.com/linsalrob/PhiSpy#installation), [Jupyter Notebooks](https://jupyter.org/install), and [pandas](https://pandas.pydata.org/pandas-docs/stable/getting_started/install.html)\n",
    "\n",
    "<small>Note: PhiSpy does not normally use pandas, but we use it here to visualize the data!</small>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# set up the environment\n",
    "import os\n",
    "import sys\n",
    "import gzip\n",
    "from functools import reduce\n",
    "import tempfile\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "from Bio import SeqIO\n",
    "\n",
    "import PhiSpyModules"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Check the PhiSpy version\n",
    "\n",
    "We recommend at least version 4.0.3 but preferable 4.1 or higher"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PhiSpy version: 4.1rc3\n"
     ]
    }
   ],
   "source": [
    "print(\"PhiSpy version: \" + PhiSpyModules.__version__)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Define your genbank file here.\n",
    "\n",
    "You may need to set the full path to the file. You can use gzip compressed file or an uncompressed file. Obviously, you will know whether it is compressed or not, but this demonstrates how PhiSpy determines when a file is compressed.\n",
    "\n",
    "**This should be the only line you need to change to run PhiSpy completely!**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "genbankfile = \"../test_genbank_files/Yersinia_pestis_KIM.gb.gz\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Parse the file\n",
    "\n",
    "We use BioPython to parse the file, but we also add a few additional mehtods to the standard BioPython object to ease parsing. We also merge or split compound features (those with more than one location along the chromosome) to appropriately handle them.\n",
    "\n",
    "Our `record` object is an extended `SeqIO.parse` object."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 6 contigs to predict prophages on!\n"
     ]
    }
   ],
   "source": [
    "min_contig_size = 1000\n",
    "\n",
    "if PhiSpyModules.is_gzip_file(genbankfile):\n",
    "    handle = gzip.open(genbankfile, 'rt')\n",
    "else:\n",
    "    handle = open(genbankfile, 'r')\n",
    "\n",
    "record  = PhiSpyModules.SeqioFilter(filter(lambda x: len(x.seq) > min_contig_size, SeqIO.parse(handle, \"genbank\")))\n",
    "handle.close()\n",
    "\n",
    "# we check to make sure there are some contigs left to process\n",
    "ncontigs = reduce(lambda sum, element: sum + 1, record, 0)\n",
    "print(f\"There are {ncontigs} contigs to predict prophages on!\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Define the parameters that we will use\n",
    "\n",
    "These are normally provided as command line options, but for jupyter we set them here\n",
    "\n",
    "paramter | meaning | options | default value\n",
    "--- | --- | --- | --\n",
    "kmers_type | What do we count kmers with? | `all`, `codon`, `simple` | `all`\n",
    "window_size | How many consecutive ORFs to include? | an integer | 30\n",
    "record | the `Bio.SeqIO` object with all the sequences | a `Bio.SeqIO` object | `record`\n",
    "expand_slope | whether to use the square of the slope of the Shannon scores | `True` or `False` | `False`\n",
    "number | Number of consecutive genes in a region of window size that must be prophage genes | an integer | 5\n",
    "nonprophage_genegaps | The number of non phage genes betweeen prophages | an integer | 10\n",
    "quiet | Don't make additional outputs | `True` or `False` | `True`\n",
    "\n",
    "\n",
    "*Note*: You can add an additional paramter, `make_training_data` here (its actual value doesn't matter) that will append an additional column to the output for each ORF that includes a `1` (`True`) if the ORF is thought or stated to be a phage gene or `0` (`False`) otherwise."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "parameters = {\n",
    "    'kmers_type': \"all\",\n",
    "    'window_size': 30,\n",
    "    'record': record,\n",
    "    'expand_slope': False,\n",
    "    'training_set': \"data/trainSet_genericAll.txt\",\n",
    "    'randomforest_trees': 5,\n",
    "    'threads': 4,\n",
    "    'quiet': True,\n",
    "    'nonprophage_genegaps': 10,\n",
    "    'number': 5\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Generate the test data\n",
    "\n",
    "This is the step that actually does all the measurements!\n",
    "\n",
    "In this example, we convert the output to a pandas dataframe for visualization and exploration."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>orf_length_med</th>\n",
       "      <th>shannon_slope</th>\n",
       "      <th>at_skew</th>\n",
       "      <th>gc_skew</th>\n",
       "      <th>max_direction</th>\n",
       "      <th>phmms</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>-36.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.801749</td>\n",
       "      <td>0.537848</td>\n",
       "      <td>7</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>-94.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.820857</td>\n",
       "      <td>0.646716</td>\n",
       "      <td>7</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>-39.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.818783</td>\n",
       "      <td>0.691991</td>\n",
       "      <td>7</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>-37.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.795715</td>\n",
       "      <td>0.633846</td>\n",
       "      <td>7</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>-33.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.900836</td>\n",
       "      <td>0.598348</td>\n",
       "      <td>7</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   orf_length_med  shannon_slope   at_skew   gc_skew  max_direction  phmms\n",
       "0           -36.0            0.0  0.801749  0.537848              7    0.0\n",
       "1           -94.5            0.0  0.820857  0.646716              7    0.0\n",
       "2           -39.0            0.0  0.818783  0.691991              7    0.0\n",
       "3           -37.5            0.0  0.795715  0.633846              7    0.0\n",
       "4           -33.0            0.0  0.900836  0.598348              7    0.0"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "parameters['test_data'] = PhiSpyModules.measure_features(**parameters)\n",
    "# note that if you include make_training_data you will need to add an \"is_phage\" column here\n",
    "test_df = pd.DataFrame(parameters['test_data'], columns = ['orf_length_med', 'shannon_slope', 'at_skew',\n",
    "                                             'gc_skew', 'max_direction', 'phmms'])\n",
    "test_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Run the random forest\n",
    "\n",
    "Here we run the random forest to identify the phages, and combine that into our initial table as the `rank` column."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene id</th>\n",
       "      <th>function</th>\n",
       "      <th>contig</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>position</th>\n",
       "      <th>rank</th>\n",
       "      <th>my status</th>\n",
       "      <th>pp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>[2:368](+)</td>\n",
       "      <td>Mobile element protein</td>\n",
       "      <td>NC_004839</td>\n",
       "      <td>3</td>\n",
       "      <td>368</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>[664:1033](+)</td>\n",
       "      <td>Plasmid conjugative transfer endonuclease</td>\n",
       "      <td>NC_004839</td>\n",
       "      <td>665</td>\n",
       "      <td>1033</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>[1170:1425](+)</td>\n",
       "      <td>Replication regulatory protein repA2 (Protein ...</td>\n",
       "      <td>NC_004839</td>\n",
       "      <td>1171</td>\n",
       "      <td>1425</td>\n",
       "      <td>2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>[1467:1587](-)</td>\n",
       "      <td>FIG01222423: hypothetical protein</td>\n",
       "      <td>NC_004839</td>\n",
       "      <td>1587</td>\n",
       "      <td>1468</td>\n",
       "      <td>3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>[1733:2600](+)</td>\n",
       "      <td>DNA replication protein</td>\n",
       "      <td>NC_004839</td>\n",
       "      <td>1734</td>\n",
       "      <td>2600</td>\n",
       "      <td>4</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          gene id                                           function  \\\n",
       "0      [2:368](+)                             Mobile element protein   \n",
       "1   [664:1033](+)          Plasmid conjugative transfer endonuclease   \n",
       "2  [1170:1425](+)  Replication regulatory protein repA2 (Protein ...   \n",
       "3  [1467:1587](-)                  FIG01222423: hypothetical protein   \n",
       "4  [1733:2600](+)                            DNA replication protein   \n",
       "\n",
       "      contig  start  stop  position  rank  my status   pp  \n",
       "0  NC_004839      3   368         0   0.0          0  1.0  \n",
       "1  NC_004839    665  1033         1   0.0          0  0.0  \n",
       "2  NC_004839   1171  1425         2   0.0          0  0.0  \n",
       "3  NC_004839   1587  1468         3   0.0          0  0.5  \n",
       "4  NC_004839   1734  2600         4   0.0          0  0.0  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "parameters['rfdata'] = PhiSpyModules.call_randomforest(**parameters)\n",
    "parameters['initial_tbl'] = PhiSpyModules.make_initial_tbl(**parameters)\n",
    "parameters['output_dir'] = tempfile.mkdtemp()\n",
    "initial_table_df = pd.DataFrame(parameters['initial_tbl'], columns = ['gene id', 'function', 'contig', 'start', 'stop', 'position', 'rank', 'my status', 'pp'])\n",
    "initial_table_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Refine the predictions\n",
    "\n",
    "Finally, we refine the predictions from the random forest and other metrics, and then predict the *att* sequences."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>num genes</th>\n",
       "      <th>att</th>\n",
       "      <th>atts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_004839</td>\n",
       "      <td>3</td>\n",
       "      <td>14951</td>\n",
       "      <td>9</td>\n",
       "      <td>[1359, 1372, 14444, 14457, GACTCAGACAGAA, GACT...</td>\n",
       "      <td>1359\\t1372\\t14444\\t14457\\tGACTCAGACAGAA\\tGACTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_004839</td>\n",
       "      <td>61369</td>\n",
       "      <td>70446</td>\n",
       "      <td>6</td>\n",
       "      <td>[60751, 60763, 67104, 67116, AAATAAAAAAAT, AAA...</td>\n",
       "      <td>60751\\t60763\\t67104\\t67116\\tAAATAAAAAAAT\\tAAAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_004838</td>\n",
       "      <td>34660</td>\n",
       "      <td>60134</td>\n",
       "      <td>13</td>\n",
       "      <td>[37969, 37982, 56736, 56749, GCCTGAAGATCCG, CG...</td>\n",
       "      <td>37969\\t37982\\t56736\\t56749\\tGCCTGAAGATCCG\\tCGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_004835</td>\n",
       "      <td>87</td>\n",
       "      <td>13676</td>\n",
       "      <td>10</td>\n",
       "      <td>[5186, 5200, 14292, 14306, GCGCTTTCTGCCGC, GCG...</td>\n",
       "      <td>5186\\t5200\\t14292\\t14306\\tGCGCTTTCTGCCGC\\tGCGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NC_004835</td>\n",
       "      <td>74658</td>\n",
       "      <td>88343</td>\n",
       "      <td>5</td>\n",
       "      <td>[75243, 75255, 89094, 89106, CATTCGGGTATA, TAT...</td>\n",
       "      <td>75243\\t75255\\t89094\\t89106\\tCATTCGGGTATA\\tTATA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NC_004836</td>\n",
       "      <td>48761</td>\n",
       "      <td>70502</td>\n",
       "      <td>12</td>\n",
       "      <td>[49151, 49249, 67536, 67634, CGCCAGACATTCACGAC...</td>\n",
       "      <td>49151\\t49249\\t67536\\t67634\\tCGCCAGACATTCACGACT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>NC_004088</td>\n",
       "      <td>502264</td>\n",
       "      <td>548451</td>\n",
       "      <td>37</td>\n",
       "      <td>[501155, 501166, 546953, 546964, TTTGTTGATGA, ...</td>\n",
       "      <td>501155\\t501166\\t546953\\t546964\\tTTTGTTGATGA\\tT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>NC_004088</td>\n",
       "      <td>2410815</td>\n",
       "      <td>2456467</td>\n",
       "      <td>36</td>\n",
       "      <td>[2410710, 2410722, 2453520, 2453532, CGGCAGATT...</td>\n",
       "      <td>2410710\\t2410722\\t2453520\\t2453532\\tCGGCAGATTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>NC_004088</td>\n",
       "      <td>2691227</td>\n",
       "      <td>2701202</td>\n",
       "      <td>6</td>\n",
       "      <td>[2692461, 2692474, 2701262, 2701275, AAAAACAGG...</td>\n",
       "      <td>2692461\\t2692474\\t2701262\\t2701275\\tAAAAACAGGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>NC_004088</td>\n",
       "      <td>3231183</td>\n",
       "      <td>3254804</td>\n",
       "      <td>19</td>\n",
       "      <td>[3232159, 3232173, 3255793, 3255807, AAATTGGTT...</td>\n",
       "      <td>3232159\\t3232173\\t3255793\\t3255807\\tAAATTGGTTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>NC_004088</td>\n",
       "      <td>3390110</td>\n",
       "      <td>3402477</td>\n",
       "      <td>10</td>\n",
       "      <td>[3389236, 3389320, 3400301, 3400385, CCAAATGTT...</td>\n",
       "      <td>3389236\\t3389320\\t3400301\\t3400385\\tCCAAATGTTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>NC_004088</td>\n",
       "      <td>3594045</td>\n",
       "      <td>3607966</td>\n",
       "      <td>6</td>\n",
       "      <td>[3592427, 3592440, 3604564, 3604577, GTAATCGGC...</td>\n",
       "      <td>3592427\\t3592440\\t3604564\\t3604577\\tGTAATCGGCT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       contig    start     stop num genes  \\\n",
       "1   NC_004839        3    14951         9   \n",
       "2   NC_004839    61369    70446         6   \n",
       "3   NC_004838    34660    60134        13   \n",
       "4   NC_004835       87    13676        10   \n",
       "5   NC_004835    74658    88343         5   \n",
       "6   NC_004836    48761    70502        12   \n",
       "7   NC_004088   502264   548451        37   \n",
       "8   NC_004088  2410815  2456467        36   \n",
       "9   NC_004088  2691227  2701202         6   \n",
       "10  NC_004088  3231183  3254804        19   \n",
       "11  NC_004088  3390110  3402477        10   \n",
       "12  NC_004088  3594045  3607966         6   \n",
       "\n",
       "                                                  att  \\\n",
       "1   [1359, 1372, 14444, 14457, GACTCAGACAGAA, GACT...   \n",
       "2   [60751, 60763, 67104, 67116, AAATAAAAAAAT, AAA...   \n",
       "3   [37969, 37982, 56736, 56749, GCCTGAAGATCCG, CG...   \n",
       "4   [5186, 5200, 14292, 14306, GCGCTTTCTGCCGC, GCG...   \n",
       "5   [75243, 75255, 89094, 89106, CATTCGGGTATA, TAT...   \n",
       "6   [49151, 49249, 67536, 67634, CGCCAGACATTCACGAC...   \n",
       "7   [501155, 501166, 546953, 546964, TTTGTTGATGA, ...   \n",
       "8   [2410710, 2410722, 2453520, 2453532, CGGCAGATT...   \n",
       "9   [2692461, 2692474, 2701262, 2701275, AAAAACAGG...   \n",
       "10  [3232159, 3232173, 3255793, 3255807, AAATTGGTT...   \n",
       "11  [3389236, 3389320, 3400301, 3400385, CCAAATGTT...   \n",
       "12  [3592427, 3592440, 3604564, 3604577, GTAATCGGC...   \n",
       "\n",
       "                                                 atts  \n",
       "1   1359\\t1372\\t14444\\t14457\\tGACTCAGACAGAA\\tGACTC...  \n",
       "2   60751\\t60763\\t67104\\t67116\\tAAATAAAAAAAT\\tAAAT...  \n",
       "3   37969\\t37982\\t56736\\t56749\\tGCCTGAAGATCCG\\tCGG...  \n",
       "4   5186\\t5200\\t14292\\t14306\\tGCGCTTTCTGCCGC\\tGCGG...  \n",
       "5   75243\\t75255\\t89094\\t89106\\tCATTCGGGTATA\\tTATA...  \n",
       "6   49151\\t49249\\t67536\\t67634\\tCGCCAGACATTCACGACT...  \n",
       "7   501155\\t501166\\t546953\\t546964\\tTTTGTTGATGA\\tT...  \n",
       "8   2410710\\t2410722\\t2453520\\t2453532\\tCGGCAGATTG...  \n",
       "9   2692461\\t2692474\\t2701262\\t2701275\\tAAAAACAGGC...  \n",
       "10  3232159\\t3232173\\t3255793\\t3255807\\tAAATTGGTTT...  \n",
       "11  3389236\\t3389320\\t3400301\\t3400385\\tCCAAATGTTA...  \n",
       "12  3592427\\t3592440\\t3604564\\t3604577\\tGTAATCGGCT...  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "parameters['pp'] = PhiSpyModules.fixing_start_end(**parameters)\n",
    "pp_df = pd.DataFrame.from_dict(parameters['pp']).transpose()\n",
    "pp_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Our `pp_df` data frame has our final prophage predictions for this genome! "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
